### ISSUE TITLE
Variable hook remains after `Variable.clone()`

### ISSUE THREAD

**Issue by yuandong-tian** [pytorch/pytorch#2601]
Here is one example. 

```
import torch.nn as nn
import torch
from torch.autograd import Variable

a = Variable(torch.FloatTensor(2, 2), requires_grad=True)
b = Variable(torch.FloatTensor(2, 2), requires_grad=True)
c = Variable(torch.FloatTensor(2, 2), requires_grad=True)

d = a * b
e = a.clone() * c

def hw_hook(s, grad):
      print(s)

a.register_hook(lambda grad: hw_hook("a1", grad))

d.backward(Variable(torch.FloatTensor(2, 2).fill_(1.0)))
e.backward(Variable(torch.FloatTensor(2, 2).fill_(1.0)))
```

It outputs two `a1` rather than one `a1`.

**Comment by apaszke** [comment]
This is expected. You can think of `.clone()` as doing `+ 0` - the output is a result of a computation that involves `a`. So both `e.backward(...)` and `d.backward(...)` will propagate gradients back to `a`.

**Comment by yuandong-tian** [comment]
@apaszke Then how to clone a Variable as if it is recomputed by its precedents again, but does not involve extra computation? I still want the gradients back-propagated. For example:

```
c = a + b
d = a + b
e = 2 * c + d * d
```
will give two Variables `c` and `d`. I can add a hook on `c` to change its gradient, but not on `d`.

**Comment by apaszke** [comment]
Do you want to modify the grad, or stop it altogether? Here are two possibilities:
For stopping:
```
c = a + b
d = c.detach()
# all code remains the same
```
For modifying:
```
val = a + b
# These two lines are 0-cost ops, but they do insert nodes in the graph,
# so c and d can have independent hooks
c = val.view_as(val)
d = val.view_as(val) 
# all code remains the same
```
